### ISSUE TITLE
Autocompleting non-apparent symbols

### ISSUE THREAD

**Issue by sorawee** [greghendershott/racket-mode#485]
For example, for the program:

```
#lang racket

(struct foo (bar baz))

foo|
```

where `|` is the current point. It should be possible to find autocomplete candidates `foo-bar`, `foo?`, etc.

One possible implementation is at https://github.com/yjqww6/drcomplete/blob/master/drcomplete-user-defined/private/main.rkt. Essentially, they fully expand the code and collect identifiers in binding positions. Lastly, they filter out identifiers that have macro scope.

Another possible implementation is my https://github.com/sorawee/fishy-completion. It is similar to DrComplete above, but also make sure that the autocompleted identifier will be in-scope. However, it is more expensive, and this means that the autocompletion must take `(point)` into account, so it might not fit nicely to the existing workflow in Racket Mode.

**Comment by greghendershott** [comment]
If `foo-bar` is imported (e.g. some other file does `(provide (struct-out foo))`) then we're good.

As for the `struct` being defined in the module, it looks like Racket Mode currently isn't aware of `syncheck:add-definition-target`, which does supply `foo-bar`. I'm not sure why. Maybe it was added in a more-recent Racket. More simply, I probably just overlooked it because its [documentation](https://docs.racket-lang.org/drracket-tools/Accessing_Check_Syntax_Programmatically.html#%28meth._%28%28%28lib._drracket%2Fcheck-syntax..rkt%29._syncheck-annotations~3c~25~3e%29._syncheck~3aadd-definition-target%29%29) has no explanatory text and I didn't notice it offered useful information not otherwise available from other annotations.

I'll update this. Thanks for pointing it out!

**Comment by greghendershott** [comment]
Restricting completion candidates by scope is definitely on my to-do list.

I think so far it's just felt like a high cost:benefit ratio.

On the cost side, I'd need to take time to figure out a reasonable data-structure. It seems obvious it should be some sort of tree, but also it needs to be easy to serialize from Racket to Emacs Lisp and reasonably fast to access in the latter, and should try to be space-efficient. This probably isn't terribly difficult, but OTOH...

On the benefit side, having excess completion candidates hasn't been too big a deal for me, in my own real-world use.

But worth doing someday.

**Comment by greghendershott** [comment]
Ha. Not only does the [doc for `syncheck:add-definition-target`](https://docs.racket-lang.org/drracket-tools/Accessing_Check_Syntax_Programmatically.html#%28meth._%28%28%28lib._drracket%2Fcheck-syntax..rkt%29._syncheck-annotations~3c~25~3e%29._syncheck~3aadd-definition-target%29%29) lack explanatory text, the bluebox signature seems wrong. It is not `(syncheck:add-definition-target source-obj start finish style-name)`. Instead it seems to be [this](https://github.com/racket/drracket/blob/master/drracket-tool-lib/drracket/private/syncheck/traversals.rkt#L1349-L1354). Anyway, in that the `(syntax-e id)` is the interesting bit we want.

**Comment by sorawee** [comment]
Thank you!

**Comment by sorawee** [comment]
Just want to add a quick note that 

1) Robby added the documentation of `syncheck:add-definition-target`, which reads: "Called to indicate a top-level definition..."
2) which means programs that have local `struct` will still fail to autocomplete:

```
#lang racket
(let ()
  (struct foo (bar))
  ;; foo-bar can't be autocompleted here
  (void))
```

There are a couple of solutions. One is to change Check Syntax to provide this information somehow. Another is to have Racket Mode walk through the expanded syntax as I described in the first comment. 

If you are OK with the latter solution, I can help implementing it and submit a PR.

**Comment by greghendershott** [comment]
Thank you for following up on this.

I'm on the fence. Not sure what to do. If anything.

- As for completion "completeness", i.e. does it offer **all** candidates: Ideally, of course, yes.

    OTOH practically it probably matters less for say local struct identifiers, because they're local. i.e. There's less value. Less typing it can help you avoid. Less "discovery" since the struct definition is in the same file and likely even on the same screen above.

    So I think this is a nice-to-have. Worth doing if it's not much work to code, maintain, and debug (including breaking other features).

- As for completion "correctness", i.e. does it off **only** candidates that would be valid in the scope at point: Ideally, of course yes.

    OTOH I feel like there are two gotchas here.

    1. It's definitely significantly more work to implement, maintain, debug.  I mean overall, not just in the code to find all of them.

    2. It probably sets an expectation that demands even more features -- starts to beg a whole "fill in the blanks with multiple choices" IDE, and more such features.

    So I feel like this is a bright line. Someday do we step over it?  Maybe. I'm not saying never. But deliberately and with a fuller plan about how far to go.

---

TL;DR: I think that's a long-winded way of saying, I'm inclined to do nothing, or, do something like the super-simple approach of DrRcomplete you mentioned in your original comment above. i.e. "Completion candidate that are complete, just maybe _too_ complete". :smile: I think that principle is still OK, for the foreseeable future.

Anyway that's my first reaction, I think?

**Comment by greghendershott** [comment]
p.s. And so, if you wanted to prepare a PR in the spirit of DrComplete's approach, I'd definitely want to merge it.

OTOH if you feel invested in your better approach, I'd totally understand if you don't want to make such a PR. Maybe in that case I'd do something on my own in the spirit of DrComplete's approach.

**Comment by sorawee** [comment]
Oh, yes, I did mean to implement DrComplete's approach.

**Comment by greghendershott** [comment]
Great. I'd be happy to take that PR.
